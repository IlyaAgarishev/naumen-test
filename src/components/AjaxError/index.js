import React from "react";
import react, { useState } from "react";
import styles from "./index.module.css";

const AjaxError = ({ ajaxError }) => {
  const [click, setClick] = useState(false);
  return (
    <div className={styles.ajaxError}>
      <div className={styles.basicDescription}>
        <span style={{ color: "red" }}>Error occurred.</span> Try to reload the
        page or check the internet connection :(
      </div>
      <div>
        <div className={styles.moreInfo} onClick={() => setClick(!click)}>
          More Info
        </div>
        {click ? <div className={styles.error}>{ajaxError.error}</div> : null}
      </div>
    </div>
  );
};

export default AjaxError;
