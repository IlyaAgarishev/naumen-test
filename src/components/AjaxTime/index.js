import React, { useContext } from "react";
import styles from "./index.module.css";
import PropTypes from "prop-types";

const AjaxTime = ({ ajaxTime }) => {
  return (
    <div className={styles.ajaxTime}>
      <div className={styles.ajaxTimeText}>Time of request:</div>
      <div className={styles.ajaxTimeNumber}>{ajaxTime} ms</div>
    </div>
  );
};

AjaxTime.propTypes = {
  ajaxTime: PropTypes.number.isRequired
};

export default AjaxTime;
