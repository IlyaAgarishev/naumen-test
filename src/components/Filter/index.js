import React from "react";
import styles from "./index.module.css";
import PropTypes from "prop-types";
import { beautifyFrequentWords } from "../../utils";

const Filter = props => {
  const { data, setFilteredData, setShowFilteredData } = { ...props };
  let mostFrequentWords = beautifyFrequentWords(data);
  return (
    <div className={styles.filter}>
      <strong>Filter by most frequent words:</strong>
      <ul className={styles.list}>
        {mostFrequentWords.map((element, index) => {
          if (element)
            return (
              <li
                key={index}
                className={styles.word}
                onClick={() => {
                  let filteredData = data.filter(
                    el => el.mostFrequentWord === element
                  );
                  setFilteredData(filteredData);
                  setShowFilteredData(true);
                }}
              >
                {element}
              </li>
            );
        })}
      </ul>
    </div>
  );
};

Filter.propTypes = {
  data: PropTypes.array.isRequired,
  setFilteredData: PropTypes.func.isRequired,
  setShowFilteredData: PropTypes.func.isRequired
};

export default Filter;
