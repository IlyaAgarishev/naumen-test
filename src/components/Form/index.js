import React, { useRef, useEffect, useContext } from "react";
import styles from "./index.module.css";
import search from "../../img/search.svg";
import PropTypes from "prop-types";
import { ajaxGetRequest, addRequest } from "../../requestFunctions";
import { MyContext } from "../../context";

const Form = props => {
  const { setData, value, setValue, requests, setAjaxError, setRequests } = {
    ...props
  };
  const darkTheme = useContext(MyContext);
  const textInput = useRef(null);

  useEffect(() => {
    textInput.current.value = value;
    if (
      navigator.userAgent.match(
        /Android|BlackBerry|iPhone|iPad|iPod|Opera Mini|IEMobile/i
      ) != null
    ) {
      return;
    }
    textInput.current.focus();
  }, [value]);

  return (
    <form
      className={styles.submitForm}
      onSubmit={e => {
        e.preventDefault();
        if (textInput.current.value === requests[0]) {
          e.preventDefault();
          return;
        }
        ajaxGetRequest(textInput.current.value)
          .then(data => {
            setData(data);
            addRequest(textInput.current.value, requests, setRequests);
            setAjaxError(false);
          })
          .catch(() => {
            setAjaxError(true);
          });
        setValue(textInput.current.value);
      }}
    >
      <input
        required
        type="text"
        ref={textInput}
        className={
          darkTheme
            ? [styles.textInput, styles.textInputDark].join(" ")
            : styles.textInput
        }
      />
      <button className={styles.submitButton} type="submit">
        <img className={styles.searchIcon} src={search} alt="" />
      </button>
    </form>
  );
};

Form.propTypes = {
  setData: PropTypes.func.isRequired,
  darkTheme: PropTypes.bool.isRequired,
  value: PropTypes.string.isRequired,
  setValue: PropTypes.func.isRequired,
  requests: PropTypes.array.isRequired,
  setAjaxError: PropTypes.func.isRequired
};

export default Form;
