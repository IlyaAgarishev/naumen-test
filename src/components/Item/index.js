import React from 'react';
import styles from './index.module.css';
import PropTypes from 'prop-types';

const Item = props => (
  <li className={styles.item}>
    <a href={props.link} className={styles.title} target="_blank" rel="noopener noreferrer">
      {props.title}
    </a>
    <div
      className={props.darkTheme ? [styles.snippet, styles.snippetDark].join(' ') : styles.snippet}
    >
      {props.snippet}
    </div>
  </li>
);

Item.propTypes = {
  title: PropTypes.string.isRequired,
  snippet: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  darkTheme: PropTypes.bool.isRequired
};

export default Item;
