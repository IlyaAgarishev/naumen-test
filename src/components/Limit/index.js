import React from "react";
import styles from "./index.module.css";
import PropTypes from "prop-types";

const Limit = props => {
  const { limit, setLimit } = { ...props };
  const limitsArray = [10, 50, 100];
  return (
    <div className={styles.limit}>
      <strong>Set limit of wiki results:</strong>
      <ul className={styles.limitButtons}>
        {limitsArray.map((number, index) => {
          return (
            <li
              className={
                limit === number
                  ? [styles.limitButton, styles.limitButtonChosen].join(" ")
                  : styles.limitButton
              }
              onClick={() => setLimit(number)}
              key={index}
            >
              {number}
            </li>
          );
        })}
      </ul>
    </div>
  );
};

Limit.propTypes = {
  limit: PropTypes.number.isRequired,
  setLimit: PropTypes.func.isRequired
};

export default Limit;
