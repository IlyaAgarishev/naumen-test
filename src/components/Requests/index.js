import React from 'react';
import styles from './index.module.css';
import search from '../../img/search.svg';
import PropTypes from 'prop-types';

const Requests = props => (
  <div className={styles.requests}>
    {props.requests.map((element, index) => {
      return (
        <div className={styles.request} key={index}>
          {element}
        </div>
      );
    })}
  </div>
);

Requests.propTypes = {
  requests: PropTypes.array.isRequired
};

export default Requests;
