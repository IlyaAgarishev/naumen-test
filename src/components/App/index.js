import React, { useState } from "react";
import burger from "../../img/burger.svg";
import styles from "./index.module.css";
import Form from "../Form";
import Item from "../Item";
import Toggle from "../Toggle";
import Requests from "../Requests";
import AjaxError from "../AjaxError";
import NoDataFound from "../NoDataFound";
import { MyContext } from "../../context";
import AjaxTime from "../AjaxTime";
import Filter from "../Filter";
import Limit from "../Limit";
import { sideBarStyle } from "../../utils";

const App = () => {
  const [requests, setRequests] = useState([]);
  const [darkTheme, setDarkTheme] = useState(false);
  const [value, setValue] = useState("");
  const [ajaxError, setAjaxError] = useState({
    error: "No error",
    status: false
  });
  const [data, setData] = useState([]);
  const [filteredData, setFilteredData] = useState([]);
  const [showFilteredData, setShowFilteredData] = useState(false);
  const [ajaxTime, setAjaxTime] = useState(0);
  const [dataNotFound, setDataNotFound] = useState();
  const [openStuff, setOpenStuff] = useState("start");
  const [limit, setLimit] = useState(10);

  const wikiResults = () => {
    if (dataNotFound) {
      return <NoDataFound />;
    } else {
      return (
        <ul className={styles.items}>
          {showFilteredData
            ? filteredData.map((element, index) => {
                return (
                  <Item
                    title={element.title}
                    snippet={element.snippet}
                    link={element.link}
                    key={index}
                  />
                );
              })
            : data.map((element, index) => {
                return (
                  <Item
                    title={element.title}
                    snippet={element.snippet}
                    link={element.link}
                    key={index}
                  />
                );
              })}
        </ul>
      );
    }
  };

  darkTheme
    ? (document.body.style.background = "#05263f")
    : (document.body.style.background = "white");

  return (
    <MyContext.Provider value={darkTheme}>
      <div className={styles.app}>
        <img
          src={burger}
          alt="burger"
          className={styles.burger}
          onClick={() => {
            setOpenStuff("open");
          }}
        />
        <div className={sideBarStyle(openStuff, styles, darkTheme)}>
          <div
            className={styles.closeBtn}
            onClick={() => setOpenStuff("close")}
          >
            X
          </div>
          <AjaxTime ajaxTime={ajaxTime} />
          <Limit limit={limit} setLimit={setLimit} />
          <Filter
            data={data}
            setFilteredData={setFilteredData}
            setShowFilteredData={setShowFilteredData}
          />
        </div>
        <Toggle setDarkTheme={setDarkTheme} />

        <div className={styles.logo}>wiki search</div>
        <Form
          setData={setData}
          value={value}
          setValue={setValue}
          requests={requests}
          setAjaxError={setAjaxError}
          setRequests={setRequests}
          setAjaxTime={setAjaxTime}
          setShowFilteredData={setShowFilteredData}
          setDataNotFound={setDataNotFound}
          setOpenStuff={setOpenStuff}
          limit={limit}
        />
        <Requests requests={requests} setValue={setValue} />
        {data.length ? (
          <div className={styles.resultsLength}>{data.length} results</div>
        ) : null}
        {ajaxError.status ? <AjaxError ajaxError={ajaxError} /> : wikiResults()}
      </div>
    </MyContext.Provider>
  );
};

export default App;
