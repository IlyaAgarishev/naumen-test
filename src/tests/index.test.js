import React from "react";
import { mount, render, shallow } from "enzyme";
import App from "../components/App";
import Item from "../components/Item";
import Toggle from "../components/Toggle";
import Form from "../components/Form";
import Requests from "../components/Requests";
import AjaxTime from "../components/AjaxTime";
import randomWords from "random-words";
import {
  beautifyResponseText,
  ajaxGetRequest,
  getRequestsStringPreview,
  addRequest,
  findMostFrequentWord,
  beautifyFrequentWords
} from "../utils";
import AjaxError from "../components/AjaxError";
import Filter from "../components/Filter";
import Limit from "../components/Limit";

// Functions testing

test("beautifyFrequentWords", () => {
  const data = [
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_1",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_2",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_3",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_3",
      link: "www.test.com"
    }
  ];
  expect(Array.isArray(beautifyFrequentWords(data))).toBe(true);
  expect(beautifyFrequentWords(data)).toEqual([
    data[0].mostFrequentWord,
    data[1].mostFrequentWord,
    data[2].mostFrequentWord
  ]);
});

test("findMostFrequentWord", () => {
  const data = "Wow Wow How How how";
  expect(typeof findMostFrequentWord(data)).toBe("string");
  expect(findMostFrequentWord(data)).toBe("how");
});

test("beautifyResponseText returns right array", () => {
  const data = [
    "Test_0",
    ["Test_1"],
    ["Wow Wow How How how Go bow bow now"],
    ["www.test.com"]
  ];
  expect(Array.isArray(beautifyResponseText(data))).toBe(true);
  expect(Object.keys(beautifyResponseText(data)[0]).length).toBe(4);
  expect(beautifyResponseText(data)).toEqual([
    {
      title: data[1][0],
      snippet: data[2][0],
      link: data[3][0],
      mostFrequentWord: findMostFrequentWord(data[2][0])
    }
  ]);
});

test("ajaxGetRequest returns right array", () => {
  const limit = [10, 50, 100];
  const randomLimit = limit[Math.floor(Math.random() * limit.length)];
  return ajaxGetRequest(randomWords()).then(data => {
    expect(data.length).toBeLessThanOrEqual(randomLimit);
    expect(Array.isArray(data)).toBe(true);
  });
});

test("ajaxGetRequest - length of objects of ajax returned array equals 4 and items in every object are strings", () => {
  return ajaxGetRequest(randomWords()).then(data => {
    for (let i = 0; i < data.length; i++) {
      expect(Object.keys(data[i]).length).toBe(4);
      Object.keys(data[i]).forEach(item => {
        expect(typeof item).toBe("string");
      });
    }
  });
});

test("getRequestsStringPreview is a string and has correct length", () => {
  const request = "request";
  const requests = ["one", "two", "three"];
  const compressedString = request + requests.join("");
  expect(typeof getRequestsStringPreview(request, requests)).toBe("string");
  expect(getRequestsStringPreview(request, requests).length).toBe(
    compressedString.length
  );
});

test("addRequest returns array and sum of lengths array elements <= 60 ", () => {
  const request = "request";
  const requests = [
    "oneoneoneoneoneoneoneoneoneoneoneoneoneone",
    "twotwotwotwotwotwotwotwotwotwotwotwotwotwotwo",
    "threethreethreethreethreethreiethreethree"
  ];
  addRequest(request, requests, data => {
    expect(Array.isArray(data)).toBe(true);
    expect(data.join("").length < 60).toBe(true);
  });
});

// Components testing

test("Item renders props correctly", () => {
  const props = {
    title: "title",
    snippet: "snippet",
    link: "link"
  };
  const component = mount(
    <Item title={props.title} snippet={props.snippet} link={props.link} />
  );
  expect(component.find(".title").text()).toBe(props.title);
  expect(component.find(".snippet").text()).toBe(props.snippet);
  expect(component.find(".title").props()).toHaveProperty("href", props.link);
});

test("Requests renders props correctly", () => {
  const func = jest.fn();
  const props = {
    requests: ["wow", "how"],
    setValue: func
  };
  const component = mount(
    <Requests requests={props.requests} setValue={props.setValue} />
  );
  for (let index = 0; index < props.requests.length; index++) {
    expect(
      component
        .find(".request")
        .at(index)
        .text()
    ).toBe(props.requests[index]);
  }
});

test("AjaxTime renders props correctly", () => {
  const time = 10;
  const component = mount(<AjaxTime ajaxTime={time} />);
  expect(component.find(".ajaxTimeNumber").text()).toBe(
    `${time.toFixed(5)} ms`
  );
});

test("Filter renders props correctly", () => {
  const data = [
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_1",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_2",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_3",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_3",
      link: "www.test.com"
    }
  ];
  const component = mount(
    <Filter
      data={data}
      setFilteredData={jest.fn()}
      setShowFilteredData={jest.fn()}
    />
  );
  let mostFrequentWords = beautifyFrequentWords(data);

  for (let index = 0; index < mostFrequentWords.length; index++) {
    expect(
      component
        .find(".word")
        .at(index)
        .text()
    ).toBe(mostFrequentWords[index]);
  }
});

test("Limit renders props correctly", () => {
  const component = mount(<Limit limit={10} setLimit={jest.fn()} />);
  const limitsArray = ["10", "50", "100"];
  for (let index = 0; index < limitsArray.length; index++) {
    expect(
      component
        .find(".limitButton")
        .at(index)
        .text()
    ).toBe(limitsArray[index]);
  }
});

// Snapshots

test("snapshots", () => {
  const data = [
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_1",
      link: "www.test.com"
    },
    {
      title: "title",
      snippet: "snippet",
      mostFrequentWord: "word_2",
      link: "www.test.com"
    }
  ];
  expect(shallow(<App />)).toMatchSnapshot();
  expect(
    shallow(<Item title={"title"} snippet={"snippet"} link={"link"} />)
  ).toMatchSnapshot();
  expect(shallow(<Toggle setDarkTheme={jest.fn()} />)).toMatchSnapshot();
  expect(
    shallow(
      <Form
        setData={jest.fn()}
        value={"Title"}
        setValue={jest.fn()}
        requests={["wow", "how"]}
        setAjaxError={jest.fn()}
        setRequests={jest.fn()}
        setAjaxTime={jest.fn()}
        setShowFilteredData={jest.fn()}
        setDataNotFound={jest.fn()}
        setOpenStuff={jest.fn()}
        limit={10}
      />
    )
  ).toMatchSnapshot();
  expect(
    shallow(<Requests requests={["wow", "how"]} setValue={jest.fn()} />)
  ).toMatchSnapshot();
  expect(shallow(<AjaxTime ajaxTime={10} />)).toMatchSnapshot();
  expect(
    shallow(
      <AjaxError
        ajaxError={{
          error: "No error",
          status: false
        }}
      />
    )
  ).toMatchSnapshot();
  expect(
    shallow(
      <Filter
        data={data}
        setFilteredData={jest.fn()}
        setShowFilteredData={jest.fn()}
      />
    )
  ).toMatchSnapshot();
  expect(shallow(<Limit limit={10} setLimit={jest.fn()} />)).toMatchSnapshot();
});
