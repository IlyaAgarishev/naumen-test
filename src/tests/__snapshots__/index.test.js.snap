// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests renders props correctly 1`] = `
<Requests
  requests={
    Array [
      "wow",
      "how",
    ]
  }
  setValue={[MockFunction]}
>
  <div
    className="requests"
  >
    <div
      className="request"
      key="0"
      onClick={[Function]}
    >
      wow
    </div>
    <div
      className="request"
      key="1"
      onClick={[Function]}
    >
      how
    </div>
  </div>
</Requests>
`;

exports[`snapshots 1`] = `
<ContextProvider
  value={false}
>
  <div
    className="app"
  >
    <AjaxTime
      ajaxTime={0}
    />
    <ToggleTheme
      setDarkTheme={[Function]}
    />
    <div
      className="logo"
    >
      wiki search
    </div>
    <Form
      requests={Array []}
      setAjaxError={[Function]}
      setAjaxTime={[Function]}
      setData={[Function]}
      setRequests={[Function]}
      setValue={[Function]}
      value=""
    />
    <Requests
      requests={Array []}
      setValue={[Function]}
    />
  </div>
</ContextProvider>
`;

exports[`snapshots 2`] = `
<li
  className="item"
>
  <a
    className="title"
    rel="noopener noreferrer"
    target="_blank"
  />
  <div
    className="snippet"
  />
</li>
`;

exports[`snapshots 3`] = `
<label
  className="toggleTheme"
>
  <Toggle
    defaultChecked={false}
    icons={false}
    onChange={[Function]}
  />
</label>
`;

exports[`snapshots 4`] = `
<form
  className="submitForm"
  onSubmit={[Function]}
>
  <input
    className="textInput"
    required={true}
    type="text"
  />
  <button
    className="submitButton"
    type="submit"
  >
    <img
      alt=""
      className="searchIcon"
      src="search.svg"
    />
  </button>
</form>
`;

exports[`snapshots 5`] = `
<div
  className="requests"
>
  <div
    className="request"
    key="0"
    onClick={[Function]}
  >
    wow
  </div>
  <div
    className="request"
    key="1"
    onClick={[Function]}
  >
    how
  </div>
</div>
`;

exports[`snapshots 6`] = `
<div
  className="ajaxTime"
>
  <div
    className="ajaxTimeText"
  >
    Time of request:
  </div>
  <div
    className="ajaxTimeNumber"
  >
     ms
  </div>
</div>
`;

exports[`snapshots 7`] = `
<div
  className="ajaxError"
>
  <div
    className="basicDescription"
  >
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      Error occurred.
    </span>
     Try to reload the page or check the internet connection :(
  </div>
  <div>
    <div
      className="moreInfo"
      onClick={[Function]}
    >
      More Info
    </div>
  </div>
</div>
`;
